generate/hcl.go

```diff
@@ -277,3 +308,134 @@ func checkHCLProviders(mod *configs.Module, opt Options) (Options, error) {
+	// stop condition, if there is no module to
+	// fetch we stop
+	if len(calls) == 0 {
+		return nil
+	}
+
+	call := calls[0]
+	name := call.Name
+
+	// we check if the module is already installed
+	// or not
+	if _, ok := installedModules[name]; ok {
+		return nil
+	}
+
+	var src string = call.SourceAddr
+
+	// we check if the module is a Terraform registry module
+	// in order to get its source address from Terraform registry
+	if regMod, err := regsrc.ParseModuleSource(src); err == nil {
+		client := registry.NewClient(nil, nil)
+		// we get the list of available module versions
+		resp, err := client.ModuleVersions(regMod)
+		if err != nil {
+			return fmt.Errorf("unable to get module versions: %w", err)
+		}
+
+		if len(resp.Modules) < 1 {
+			return fmt.Errorf("unable to find suitable versions")
+		}
+		meta := resp.Modules[0]
+
+		var (
+			latest *version.Version
+			// match holds the version matching the
+			// source constraints set in the module call
+			match *version.Version
+		)
+		for _, vers := range meta.Versions {
+			v, err := version.NewVersion(vers.Version)
+			if err != nil {
+				return fmt.Errorf("unable to create version from string: %w", err)
+			}
+
+			if latest == nil || v.GreaterThan(latest) {
+				latest = v
+			}
+
+			if call.Version.Required.Check(v) {
+				if match == nil || v.GreaterThan(match) {
+					match = v
+				}
+			}
+		}
+		// we finally get the module location, it will return
+		// a string `go-getter` compliant
+		src, err = client.ModuleLocation(regMod, match.String())
+		if err != nil {
+			return fmt.Errorf("unable to fetch module location: %w", err)
+		}
+	}
+
+	// since go-getter does not support yet in-memory fs,
+	// we need to initialize the parser using actual fs
+	// https://github.com/hashicorp/go-getter/issues/83
+	pars := configs.NewParser(nil)
+
+	// we check if the module is a local one by checking
+	// its prefix "./", "../", etc.
+	var isLocal bool
+	for _, prefix := range localSourcePrefixes {
+		if strings.HasPrefix(src, prefix) {
+			isLocal = true
+		}
+	}
+
+	var (
+		m     *configs.Module
+		diags hcl.Diagnostics
+	)
+
+	// the module is not a local one or a Terraform registry one
+	// it should be handle by `go-getter`
+	if !isLocal {
+		dst := path.Join(cachePath, name)
+		// TODO: we should add a logic to invalidate
+		// the cache
+		if _, err := os.Stat(dst); os.IsNotExist(err) {
+			client := &getter.Client{
+				Src:  src,
+				Dst:  dst,
+				Pwd:  pwd,
+				Mode: getter.ClientModeDir,
+			}
+			if err := client.Get(); err != nil {
+				return fmt.Errorf("unable to get remote module: %w", err)
+			}
+		}
+		m, diags = pars.LoadConfigDir(dst)
+	} else {
+		m, diags = pars.LoadConfigDir(path.Join(pwd, src))
+	}
+	if diags.HasErrors() {
+		return fmt.Errorf("unable to load config directory: %s", diags.Error())
+	}
+
+	// fill the final map of managed resources
+	// using the config freshly loaded
+	for rk, rv := range m.ManagedResources {
+		(*mRes)[rk] = rv
+	}
+
```

Comment:
Potentially prefix it with them module if more than one or we'll have collision.